autograd_test1.py

```diff
@@ -0,0 +1,602 @@
+        return out1-out2
+
+class accuracy(operation):
+    def forward(self, pred, target):
+        y = np.argmax(pred, axis=1)
+        t = np.argmax(target, axis=1)
+        a = y == t
+        return np.array(a, 'int').sum() / len(t)
+    def backward(self, *grad_output):
+        pass
+def accuracy_np(y,t):
+    y = np.argmax(y, axis=1)
+    t = np.argmax(t, axis=1)
+    a = y == t
+    return np.array(a, 'int').sum() / len(t)
+
+class cross_entropy(operation):
+    def forward(self, pred,target):
+        out = - np.multiply(target, np.log(pred)).sum() / pred.shape[0]
+        self.input=(pred,target)
+        #self.output=(out,)
+        return out
+    def backward(self, grad_output):
+        #print(self.input[1]/self.input[0], '223123')
+        #a=-np.ones(self.input[0].shape)*(self.input[1]/self.input[0])/(self.input[0].shape[0])
+        a = -(self.input[1] / self.input[0]) / (self.input[0].shape[0])
+
+        #print(grad_output*a)
+        return grad_output*a
+
+class output(operation):
+    def forward(self, z,T):
+        Y=np.exp(z) / np.sum(np.exp(z), axis=1, keepdims=True)
+        loss= - np.multiply(T, np.log(Y)).sum() / Y.shape[0]
+        accu=accuracy_np(Y,T)
+        print('accuracy:',accu)
+        self.input=(Y,T)
+        return loss
+    def backward(self,grad_output):
+        return grad_output*(self.input[0] - self.input[1]) / self.input[0].shape[0]
+
+class ExecutionEngine(object):
+    def __init__(self):
+        pass
+
+    def _compute_dependencies(self, function):
+        dependencies = {}
+        seen = {function}
+        queue = [function]
+        while len(queue) > 0:
+            fn = queue.pop()
+            for prev_fn, arg_id in fn.previous_functions:
+
+                if prev_fn not in dependencies:
+                    dependencies[prev_fn] = [Counter() for _ in prev_fn.output_ids]
+                    #print(dependencies[prev_fn])
+                #output_idx=0
+                output_idx = prev_fn.output_ids[arg_id]
+                dependencies[prev_fn][output_idx][fn] += 1 #dependencies[prev_fn]:list of [Counter(),... ], output_idx:index,take out an Counter dict
+                if prev_fn not in seen:
+                    queue.append(prev_fn)
+                    seen.add(prev_fn)
+        return dependencies,seen  #记录了 某一个函数通过其第几个输出依赖于其后一个函数
+
+    def _free_backward_dependency(self, dependencies, prev_fn, fn, arg_id):
+        deps = dependencies[prev_fn]
+        output_idx = prev_fn.output_ids[arg_id]
+        output_deps = deps[output_idx] #Counter()
+        output_deps[fn] -= 1
+        if output_deps[fn] == 0:
+            del output_deps[fn]
+        return output_idx#返回prev_fn的为arg_id输出的下标
+
+
+    def _is_ready_for_backward(self, dependencies, function):
+        for deps in dependencies[function]: #list
+            if len(deps) > 0:
+                return False
+        return True
+
+    def run_backward(self, variable, grad):
```

Comment:
https://github.com/HIPS/autograd/blob/master/autograd/core.py#L17

The backward (and dependency sort) in autograd is very clear.
what you need is just to count the available output tensors. if all output tensors are ready, then you can do backward.